File: src/storage_system.cc
#include "endfield_base/storage_system.h"

namespace endfield_base {
StorageEndpoints StorageSystem::collectEndpoints(const SimulationState& state) {
    StorageEndpoints endpoints;
    for (const FacilityInstance& instance : state.grid.getFacilities()) {
        const FacilityDefinition* definition = state.catalog.findDefinition(instance.definitionId);
        if (definition == nullptr) {
            continue;
        }

        if (definition->category == FacilityCategory::Storage_in) {
            endpoints.inputInstanceIds.push_back(instance.instanceId);
        } else if (definition->category == FacilityCategory::Storage_out) {
            endpoints.outputInstanceIds.push_back(instance.instanceId);
        } else if (isHubFacility(*definition) || definition->category == FacilityCategory::Storage_port) {
            if (definition->maxInputs > 0) {
                endpoints.outputInstanceIds.push_back(instance.instanceId);
            }
            if (definition->maxOutputs > 0) {
                endpoints.inputInstanceIds.push_back(instance.instanceId);
            }
        } else if (usesWirelessStorage(*definition, instance)) {
            if (definition->maxOutputs > 0) {
                endpoints.outputInstanceIds.push_back(instance.instanceId);
            }
        }
    }

    return endpoints;
}
}  // namespace endfield_base

File: src/EndfieldBase.Core/Storage/StorageSystem.cs
using EndfieldBase.Core.Models;

namespace EndfieldBase.Core.Storage;

public sealed class StorageEndpoints
{
    public List<int> InputInstanceIds { get; } = [];
    public List<int> OutputInstanceIds { get; } = [];
}

public static class StorageSystem
{
    public static StorageEndpoints CollectEndpoints(SimulationState state) {
        StorageEndpoints endpoints = new();
        foreach (var instance in state.Grid.GetFacilities()) {
            var definition = state.Catalog.FindDefinition(instance.DefinitionId);
            if (definition is null) continue;

            if (definition.Category == FacilityCategory.StorageIn) {
                endpoints.InputInstanceIds.Add(instance.InstanceId);
            }
            else if (definition.Category == FacilityCategory.StorageOut) {
                endpoints.OutputInstanceIds.Add(instance.InstanceId);
            }
            else if (FacilityRules.IsHubFacility(definition) || definition.Category == FacilityCategory.StoragePort) {
                if (definition.MaxInputs > 0) endpoints.OutputInstanceIds.Add(instance.InstanceId);

                if (definition.MaxOutputs > 0) endpoints.InputInstanceIds.Add(instance.InstanceId);
            }
            else if (FacilityRules.UsesWirelessStorage(definition, instance) && definition.MaxOutputs > 0) {
                endpoints.OutputInstanceIds.Add(instance.InstanceId);
            }
        }

        return endpoints;
    }
}